src/Microsoft.IdentityModel.Tokens/JsonWebKeySet.cs

```diff
@@ -122,68 +120,100 @@ public JsonWebKeySet(string json, JsonSerializerSettings jsonSerializerSettings)
-                            }
-                        }
-                    }
+                        signingKeys.AddRange(CreateX509SecurityKeys(webKey));
 
                     if (!string.IsNullOrWhiteSpace(webKey.E) && !string.IsNullOrWhiteSpace(webKey.N))
+                        signingKeys.Add(CreateRsaSecurityKey(webKey));
+                }
+                else if (webKey.Kty.Equals(JsonWebAlgorithmsKeyTypes.EllipticCurve, StringComparison.Ordinal))
+                {
+                    signingKeys.Add(CreateECDsaSecurityKey(webKey));
+                }
+                else
+                {
+                    //kty is not 'EC' or 'RSA'
+                    LogHelper.LogInformation(LogHelper.FormatInvariant(LogMessages.IDX10809, webKey.Kty));
+                }
+            }
+
+            return signingKeys;
+        }
+
+        private IList<SecurityKey> CreateX509SecurityKeys(JsonWebKey jsonWebKey)
+        {
+            try
+            {
+                var keys = new List<SecurityKey>();
+
+                foreach (var certString in jsonWebKey.X5c)
+                {
+                    var key = new X509SecurityKey(new X509Certificate2(Convert.FromBase64String(certString)))
                     {
-                        try
-                        {
-                            SecurityKey key =
-                                 new RsaSecurityKey
-                                 (
-                                    new RSAParameters
-                                    {
-                                        Exponent = Base64UrlEncoder.DecodeBytes(webKey.E),
-                                        Modulus = Base64UrlEncoder.DecodeBytes(webKey.N),
-                                    }
-
-                                );
-                            key.KeyId = webKey.Kid;
-                            keys.Add(key);
-                        }
-                        catch (CryptographicException ex)
-                        {
-                            throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10801, webKey.E, webKey.N), ex));
-                        }
-                        catch (FormatException ex)
-                        {
-                            throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10801, webKey.E, webKey.N), ex));
-                        }
-                    }
+                        KeyId = jsonWebKey.Kid
+                    };
+
+                    keys.Add(key);
                 }
+
+                return keys;
             }
+            catch (Exception ex)
+            {
+                throw LogHelper.LogExceptionMessage(new InvalidOperationException(LogHelper.FormatInvariant(LogMessages.IDX10802, jsonWebKey.X5c[0]), ex));
```

Comment:
I'm wondering about throwing here. Suppose there is one bad key and a bunch of good ones.

Reply:
This is also in-line with the existing behavior. I'll think about the alternative solutions.

Reply:
@brentschmaltz  As this feature is planned for 5.5.0 release, we should keep the existing behavior imo.
I'm thinking of adding a flag that would control skipping invalid signing keys (or throwing). 

Users could easily opt-out by flipping the flag and we won't throw an exception whenever we see an invalid signing key.

Reply:
addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1143/commits/5f40af68917573f4fae7a80c0bb542f6ad275b5e